lib/model/unreads.dart

```diff
@@ -206,6 +206,15 @@ class Unreads extends ChangeNotifier {
     }
   }
 
+  /// Checks if stream contains strictly muted unreads,
+  /// using [ChannelStore.isTopicVisible].
+  bool hasMutedInStream(int streamId) {
```

Comment:
This will include unreads that are in muted topics. That doesn't sound like the desired behavior — and it's not what web does, based on checking just now with a channel where I have some such unreads.

I think the desired behavior is: we want to see if there are any unreads here where `isTopicVisibleInStream` is true but the stricter `isTopicVisible` isn't.

In other words:
* We show an indicator on the channel/stream when there are some unreads that we would show if you narrow to the channel. (That's what `countInStreamNarrow` / `isTopicVisibleInStream` would measure.)
* When moreover some of those unreads are messages we'd show in the combined feed, then the indicator we show is a count of those unreads. (That's `countInStream` / `isTopicVisible`.)
* So we show a dot when the first condition is true but not the second.

Can you find the code in web that controls this aspect? That'd be good for confirming what web's behavior is to verify that my description above matches it.

Then for implementing that behavior, probably cleanest is to use the existing method `countInStreamNarrow`.

Reply:
Thanks @gnprice! 

Yes, I can verify you're description:

> I think the desired behavior is: we want to see if there are any unreads here where isTopicVisibleInStream is true but the stricter isTopicVisible isn't.

I've taken a look into the web code and found that the only case that applies is when stream isMuted (or `stream.isMuted` is null) and topic policy is none. and that Equals to `isTopicVisibleInStream && !isTopicVisible`.

| topic policy | stream isMuted | isTopicVisible | isTopicVisibleInStream | isTopicVisibleInStream && !isTopicVisible | web counts as unmuted unread |
| --- | --- | --- | --- | --- | --- |
| none | true | false | true | true | true |
| muted | true | false | false | false | false |
| unmuted | true | true | true | false | false |
| followed | true | true | true | false | false |
| none | null | true | true | false | true |
| muted | null | false | false | false | false |
| unmuted | null | true | true | false | false |
| followed | null | true | true | false | false |
| none | false | true | true | false | false |
| muted | false | false | false | false | false |
| unmuted | false | true | true | false | false |
| followed | false | true | true | false | false |

You can find that in [web/src/unread.ts@UnreadTopicCounter.get_stream_count_info](https://github.com/zulip/zulip/blob/main/web/src/unread.ts#L393)

```ts
get_stream_count_info(
    stream_id: number,
    update_first_unmuted_message_id = false,
): StreamCountInfo {
    const per_stream_bucketer = this.bucketer.get(stream_id);

    if (!per_stream_bucketer) {
        return {
            unmuted_count: 0,
            muted_count: 0,
            followed_count: 0,
            stream_is_muted: false,
        };
    }

    const sub = sub_store.get(stream_id);
    let unmuted_count = 0;
    let muted_count = 0;
    let followed_count = 0;
    for (const [topic, msgs] of per_stream_bucketer) {
        const topic_count = msgs.size;

        if (user_topics.is_topic_followed(stream_id, topic)) {
            followed_count += topic_count;
        }

        if (user_topics.is_topic_unmuted_or_followed(stream_id, topic)) {
            unmuted_count += topic_count;
            if (update_first_unmuted_message_id) {
                first_unread_unmuted_message_id = Math.min(
                    first_unread_unmuted_message_id,
                    ...msgs,
                );
            }
        } else if (user_topics.is_topic_muted(stream_id, topic)) {
            muted_count += topic_count;
        } else if (sub?.is_muted) {
            muted_count += topic_count;
        } else {
            unmuted_count += topic_count;
            if (update_first_unmuted_message_id) {
                first_unread_unmuted_message_id = Math.min(
                    first_unread_unmuted_message_id,
                    ...msgs,
                );
            }
        }
    }
    const stream_count = {
        unmuted_count,
        muted_count,
        followed_count,
        stream_is_muted: sub?.is_muted ?? false,
    };
    return stream_count;
}
```

And in [web/src/stream_list.ts@update_count_in_dom](https://github.com/zulip/zulip/blob/main/web/src/stream_list.ts#L100)

```ts
export function update_count_in_dom(
    $stream_li: JQuery,
    stream_counts: StreamCountInfo,
    stream_has_any_unread_mention_messages: boolean,
    stream_has_any_unmuted_unread_mention: boolean,
    stream_has_only_muted_unread_mention: boolean,
): void {
    // The subscription_block properly excludes the topic list,
    // and it also has sensitive margins related to whether the
    // count is there or not.
    const $subscription_block = $stream_li.find(".subscription_block");

    ui_util.update_unread_mention_info_in_dom(
        $subscription_block,
        stream_has_any_unread_mention_messages,
    );

    if (stream_has_any_unmuted_unread_mention) {
        $subscription_block.addClass("has-unmuted-mentions");
        $subscription_block.removeClass("has-only-muted-mentions");
    } else {
        $subscription_block.removeClass("has-unmuted-mentions");
        if (!stream_counts.stream_is_muted && stream_has_only_muted_unread_mention) {
            $subscription_block.addClass("has-only-muted-mentions");
        } else {
            $subscription_block.removeClass("has-only-muted-mentions");
        }
    }

    // Here we set the count and compute the values of two classes:
    // .stream-with-count is used for the layout CSS to know whether
    // to leave space for the unread count, and has-unmuted-unreads is
    // used in muted streams to set the fading correctly to indicate
    // those are unread
    if (stream_counts.unmuted_count > 0 && !stream_counts.stream_is_muted) {
        // Normal stream, has unmuted unreads; display normally.
        ui_util.update_unread_count_in_dom($subscription_block, stream_counts.unmuted_count);
        $subscription_block.addClass("stream-with-count");
        $subscription_block.removeClass("has-unmuted-unreads");
        $subscription_block.removeClass("has-only-muted-unreads");
    } else if (stream_counts.unmuted_count > 0 && stream_counts.stream_is_muted) {
        // Muted stream, has unmuted unreads.
        ui_util.update_unread_count_in_dom($subscription_block, stream_counts.unmuted_count);
        $subscription_block.addClass("stream-with-count");
        $subscription_block.addClass("has-unmuted-unreads");
        $subscription_block.removeClass("has-only-muted-unreads");
    } else if (stream_counts.muted_count > 0 && stream_counts.stream_is_muted) {
        // Muted stream, only muted unreads.
        ui_util.update_unread_count_in_dom($subscription_block, stream_counts.muted_count);
        $subscription_block.addClass("stream-with-count");
        $subscription_block.removeClass("has-unmuted-unreads");
        $subscription_block.removeClass("has-only-muted-unreads");
    } else if (
        stream_counts.muted_count > 0 &&
        !stream_counts.stream_is_muted &&
        stream_has_only_muted_unread_mention
    ) {
        // Normal stream, only muted unreads, including a mention:
        // Display the mention, faded, and a faded unread count too,
        // so that we don't weirdly show the mention indication
        // without an unread count.
        ui_util.update_unread_count_in_dom($subscription_block, stream_counts.muted_count);
        $subscription_block.removeClass("has-unmuted-unreads");
        $subscription_block.addClass("stream-with-count");
        $subscription_block.addClass("has-only-muted-unreads");
    } else if (stream_counts.muted_count > 0 && !stream_counts.stream_is_muted) {
        // Normal stream, only muted unreads: display nothing. The
        // current thinking is displaying those counts with muted
        // styling would be more distracting than helpful.
        ui_util.update_unread_count_in_dom($subscription_block, 0);
        $subscription_block.removeClass("has-unmuted-unreads");
        $subscription_block.removeClass("stream-with-count");
    } else {
        // No unreads: display nothing.
        ui_util.update_unread_count_in_dom($subscription_block, 0);
        $subscription_block.removeClass("has-unmuted-unreads");
        $subscription_block.removeClass("has-only-muted-unreads");
        $subscription_block.removeClass("stream-with-count");
    }
}
```

Reply:
Cool, thanks for investigating that.

In your table, in the row for none/null, isTopicVisible should be false, not true, right? We have:
```dart
  bool isTopicVisible(int streamId, String topic) {
    switch (topicVisibilityPolicy(streamId, topic)) {
      case UserTopicVisibilityPolicy.none:
        switch (subscriptions[streamId]?.isMuted) {
          case false: return true;
          case true:  return false;
          case null:  return false; // not subscribed; treat like muted
```

so none/null behaves the same as none/true.

That's good, though, because I think that's what makes this conclusion true:

> the only case that applies is when stream isMuted (or `stream.isMuted` is null) and topic policy is none. and that Equals to `isTopicVisibleInStream && !isTopicVisible`.